File: JAVA/PartitionProblemBackTrackingRecursive/src/partitionproblembacktrackingrecursive/BackTrackingR.java
package partitionproblembacktrackingrecursive;

public class BackTrackingR {
    private final int[] num;
    private final int n;
    private int sum = 0;
    private boolean found;
    
    public BackTrackingR(int[] num){
        this.num = num;
        this.n = num.length;        
    }
    
    public void printVector(){
        System.out.print("| ");
        for (int i = 0; i < n; i++) { 
            System.out.print(num[i] + " | "); 
        } 
    } 
    
    private void getResult(){
        System.out.println("Results: " + sum);
    }
    
    public void printResult(){
        if(found){
            System.out.println("\nVector can be partitioned");       
            getResult();
        }else{
            System.out.println("\nVector can not be partitioned");
        }
    }
    
    public void printTime(long timeMillis, long timeNano, String scale){
        System.out.print("For " + n + " elements -> " + "Taken time: ");
        float print = (float) ((timeMillis == 0)? timeNano/1e9d : timeMillis/1e3d);
        if(scale == null){            
            System.out.printf("%.8f s.\n", print);
        }else{
            switch (scale) {
                case "ns": System.out.println(timeNano + " " + scale + '.');
                break;
                case "ms": System.out.println(timeMillis + " " + scale + '.');
                break;
                default: System.out.printf("%.8f s.\n", print);
                break;
            }  
        }
    }
    
    private int add(){
        for (int i = 0; i < n; i++) sum += num[i];
        return sum;
    }
    
    public boolean canPartition() {
        int su = add();
        return (su % 2 == 0);
    }
    
    public void canPartitionRecursiveAux(){
        sum /= 2;        
        found = canPartitionRecursive(n-1, sum);
    }
    
    private boolean canPartitionRecursive(int n, int s) {
        if (s <= 0) return true;

        if(n < 0) return false;

        if(s < num[n]){
            return canPartitionRecursive(n-1, s);
        }else{
            return (canPartitionRecursive(n-1, s) ||
		    canPartitionRecursive(n-1, s-num[n]));
        }
    }
}

File: CPP/PartitionProblemBackTrackingRecursive/BackTrackingR.cpp
#include "BackTrackingR.h"

BackTrackingR::BackTrackingR(int* num, int size) {
    this->num = num;
    this->n = size;
}

void BackTrackingR::printVector() {
    cout << "| ";
    for (int i = 0; i < n; i++) {
        cout << num[i] << " | ";
    }
    cout << endl;
}

void BackTrackingR::getResult() {
    cout << "Results: " << sum << endl;    
}

void BackTrackingR::printResult() {
    if(found){
        cout << "\nVector can be partitioned" << endl;
        BackTrackingR::getResult();
    }else{
        cout << "\nVector can not be partitioned" << endl;
    }
}

void BackTrackingR::printTime(long timeMS, long timeNS, string scale) {
    printf("For %d elements -> Taken time: ", n); 
    float print = (float) ((timeMS == 0)? timeNS/1e9 : timeMS/1e3);
    if(scale.empty()){
        printf("%.8f s.\n", print);
    }else{
        if(scale == "ns"){
            cout << timeNS << " " << scale << '.' << endl;
        }else if(scale == "ms"){
            cout << timeMS << " " << scale << '.' << endl;
        }else{
            printf("%.8f s.\n", print);
        }
    }
}

int BackTrackingR::add() {
    for (int i = 0; i < n; i++) sum += num[i];
        
    return sum;
}

bool BackTrackingR::canPartition() {
    int su = add();
    return (su % 2 == 0);
}

void BackTrackingR::canPartitionRecursiveAux(){
    sum /= 2;
    found = BackTrackingR::canPartitionRecursive(n-1, sum);
}

bool BackTrackingR::canPartitionRecursive(int n, int s) {
    if (s <= 0) return true;

    if(n < 0) return false;

    if(s < num[n]){
        return BackTrackingR::canPartitionRecursive(n-1, s);
    }else{
        return (BackTrackingR::canPartitionRecursive(n-1, s) |
                BackTrackingR::canPartitionRecursive(n-1, s-num[n]));
    }
        
}


